src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs

```diff
@@ -444,27 +444,14 @@ public void Release()
 
                 public LineBatch GetBatch()
                 {
-                    Exception inner = null;
-                    try
-                    {
-                        var batch = _queue.Take();
-                        if (batch.Exception == null)
-                            return batch;
-                        inner = batch.Exception;
-                    }
-                    catch (InvalidOperationException)
-                    {
-                        // This code detects when there is no more content by catching the exception thrown by _queue.Take() at the end.
-                        // If _queue.IsAddingCompleted is true, we know that this exception was the result of that specifically.
-                        // This should probably be re-engineered to not rely on exception blocks.
-                        // REVIEW: Come up with a less strange scheme for the interthread communication here, that does not
-                        // rely on exceptions and timeouts throughout the pipeline.
-                        if (_queue.IsAddingCompleted)
-                            return default(LineBatch);
-                        throw;
-                    }
-                    Contracts.AssertValue(inner);
-                    throw Contracts.ExceptDecode(inner, "Stream reading encountered exception");
+                    if (!_queue.TryTake(out LineBatch batch, millisecondsTimeout: -1))
+                        return default;
```

Comment:
> Is IsAddingCompleted the only reason TryTake will return false when given an Infinite timeout?

Yeah... well, or a Dispose.

(Related, I think there may be some additional exceptional issues happening because of disposal in fact.  I'm seeing a non-trivial number of OperationCanceledExceptions getting thrown and swallowed inside of BlockingCollection due to it being disposed of while there are still threads actively waiting to take.  While it'd be great if BC could avoid using exceptions itself for control flow in that case, it's generally not something to optimize for given you're generally not supposed to dispose of things while they're still in use.)